### ISSUE TITLE
More consistent empty set signs for default font

### ISSUE THREAD

**Issue by sitandr** [typst/typst#1528]
### Description

See discussion in #1526.

As @Enivex mentioned,

 > the correct way of dealing with this is using the cv01 font feature, as the font provides an alternative version of the symbol. Though I do not believe this feature is currently accessible in Typst.

So the idea is fixing default math font with font features instead of directly swapping the symbols for all the fonts.

### Use Case

That issue (need to use `diameter` instead of `nothing` for default math font) might be really confusing for newcomers, so it should be at least mentioned somewhere until fixed.

**Comment by Enivex** [comment]
To be clear, the character variant font features should be exposed as an option, just like stylistic sets. Then the user can enable cv1 for new computer modern math if they would like the alternate empty set symbol

**Comment by sitandr** [comment]
I'm not sure about this. After all, it seems to me that typst want to provide nice "bindings" to symbols, rather than inventing new symbol names.

So I think it would be much more user-friendly to have that feature enabled by default. Of course, if there are no major drawbacks like compatibility issues.

**Comment by Enivex** [comment]
> I'm not sure about this. After all, it seems to me that typst want to provide nice "bindings" to symbols, rather than inventing new symbol names.
> 
> So I think it would be much more user-friendly to have that feature enabled by default. Of course, if there are no major drawbacks like compatibility issues.

The feature would do something else for other fonts, so that's not a good idea

**Comment by sitandr** [comment]
Oh, I thought you are talking about enabling that feature by default *only for default math font* that has that issue. Sorry for misunderstanding.

So this is a bad way of dealing with that?

**Comment by sitandr** [comment]
Eh, that reminds me LaTex way of dealing with things: let user try finding on StackOverflow all those large preamble parts and import one billion of packages to make the writing barely bearable.

I mean, I like the way Typst uses lots of reasonable defaults, and that could be one of those (again, if there are no major downsides).

Of course, it is up to founders to decide.

**Comment by Enivex** [comment]
I don't agree with that. The reasonable default is the default for the font.

Regardless, right now the opentype font features are tied to text, and not the actual font. That would have to change first. It's already been discussed in the Discord.

**Comment by user202729** [comment]
The reason people prefer the circular one to the thin one is somewhat historical -- LaTeX's "old" math font supports the thin one as `\emptyset` and the circular one as `\varemptyset` (or `\varnothing`), and (for some reason) people prefer the "variant" version to the original version.

In the new font "only one" is supported because Unicode mapping is semantic.

Either way, setting font-default to be the default character seems to be less confusing. Imagine the user making a modification to the font in one irrelevant entry and notice after a month later that the `nothing` symbol has changed, and have no idea why.

(admittedly users making custom font is less common than user using ∅, but there's still some user getting confused)

**Comment by SkiFire13** [comment]
I've been using the `text(font: "", emptyset)` hack to work around this issue for a while, but it seems that the `main` branch version of `typst` warns against it

```
warning: unknown font family:
  ┌─ \\?\E:\Playground\test.typ:1:13
  │
1 │ $ text(font: "", emptyset) $
  │
```

Is there a better alternative that doesn't produce a warning?

**Comment by laurmaedje** [comment]
Maybe ` $ text(font: #(), emptyset) $` instead? That should have a similar effect. But it might warn as well in the future.

**Comment by MDLC01** [comment]
I don't think this was mentioned in this thread, but the modern way of differentiating between wide and narrow "empty set" symbols is using Unicode variation sequences. Essentially, both the narrow and the wide forms are valid glyphs for U+2205 EMPTY SET,[^1] but adding the U+FE00 VARIATION SELECTOR-1 character right after U+2205 indicates to the font that the "zero with long diagonal stroke overlay form" (i.e., narrow form) should be used.[^2]

New Computer Modern Math, the font used by default for math in Typst, displays U+2205 using the narrow form even without the variation selector. This is conformant to Unicode, but it means the only way to access the wide form is through the alternate means of [enabling the non-standard `cv01` feature](https://typst.app/docs/reference/text/text/#parameters-features). When the feature is enabled, New Computer Modern Math always uses the wide form, even when the variation selector is present. Ideally, New Computer Modern Math would instead use the wide form by default for U+2205 and switch to the narrow form when followed by U+FE00. @mkorje recently made the maintainer of New Computer Modern Math aware of this issue.

On our side, given that Unicode provides a way to specifically require a narrow form, we should definitely allow users to access this form as a named symbol. Since Typst 0.14.0, symbols consisting of multiple codepoints are fully supported.[^3] This means once we decide what name to use for the narrow variant, the implementation will be very easy. Discussion regarding the name should happen on [Discord](https://discord.com/channels/1054443721975922748/1277628305142452306).

In the meantime, the following code snippet can be used to access the wide variant when using New Computer Modern Math:

```typ
#let wideemptyset = text(features: ("cv01",), $emptyset$)

// Example:
$ emptyset != wideemptyset $
```

[^1]: Note that, as per the [Unicode 17.0.0 Core Specification, Section 23.4](https://www.unicode.org/versions/Unicode17.0.0/core-spec/chapter-23/#G26678), "The standardization or support of a particular variation sequence does _not_ limit the set of glyphs that can be used to represent the initial character alone." In other words, both forms are valid for U+2205 alone. It is only when followed by VS1 that the form is restricted.
[^2]: As defined in the [Mathematical Operators block chart](https://unicode.org/charts/PDF/U2200.pdf).
[^3]: Support for multi-codepoint glyphs in math implemented in #6336. Support for multi-codepoint symbols implemented in #6489.

**Comment by ponte-vecchio** [comment]
As of today, [kpfonts-otf](https://ctan.org/ctan-ann/id/54289d67-5694-01ae-383c-9cbd4b382177@ctan.org), [xcharter-math](https://ctan.org/ctan-ann/id/edcb680e-8049-b5b9-2c4e-1b66ab4d1bfd@ctan.org) and [erewhon-math](https://ctan.org/ctan-ann/id/edcb680e-8049-b5b9-2c4e-1b66ab4d1bfd@ctan.org) have added support for the use of Variation Selector `U+FE00`.

**Comment by mkorje** [comment]
The latest version of New CM has switched the emptyset glyph with the one in `cv01`. I've opened https://github.com/typst/codex/pull/132 to add a name for the narrow variant using variation selectors.

**Comment by XanderLeaDaren** [comment]
Hi there, any news on the emptyset/diameter issue?

It seems rather odd for newcomers that `diameter` is related to all `emptyset` variants…

```
$
  emptyset \

  diameter \
  
  emptyset.arrow emptyset.arrow.l emptyset.bar emptyset.circle emptyset.rev
$
```

<img width="108" height="105" alt="Image" src="https://github.com/user-attachments/assets/689f3b0f-5ad0-421c-89e6-aeab59894ed7" />

**Comment by Enivex** [comment]
It should already be fixed (that's why the issue was closed).

**Comment by laurmaedje** [comment]
I can confirm that it's fixed (but not yet released) for the font bundled in the CLI. If you're running locally, you may also need to use `--ignore-system-fonts` if you have an outdated version of New CM installed. The embedded one looks correct in the development version. On the web app, the font is not yet updated (also not for the development version since all versions share the same fonts).
